File: 2015/01/algebraic-data-types-in-c/main.hs
data Point = Point { x :: Float
                   , y :: Float
                   }

data Shape = Circle { center :: Point
                    , r :: Float
                    }
           | Rectangle { topleft :: Point
                       , bottomright :: Point
                       }

main :: IO ()
main = do
  putStrLn "Test."

File: 2015/01/algebraic-data-types-in-c/main.c
#include <stdio.h>

typedef struct {
  double x;
  double y;
} Point;

typedef enum {
  CircleTag,
  RectangleTag
} ShapeTag;

typedef struct {
  Point p;
  double r;
} Circle;

typedef struct {
  Point topleft;
  Point topright;
} Rectangle;

typedef struct {
  ShapeTag tag;
  union {
    Circle c;
    Rectangle r;
  };
} Shape;

int main() {
  puts("Test.");
  return 0;
};
